feat(client): handle quiz finish and exit

huyenltnguyen opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

This PR:
- Adds `passingGrade` to quiz
- Adds a `usePageLeave` hook to subscribe to the page close and page change events
- Updates the quiz workflow to handle the finish and exit actions
- Adds tests for the quiz challenge since its behaviors are now a little too complex to keep track of

The workflow is now as follows:
- If campers refresh, click the back button, or click on any link on the page
  - The action is blocked, and a confirm dialog would show up
- If campers have not answered all questions
  - Clicking the Finish button would show a list of unanswered questions
- If campers have answered all questions
  - Clicking the Finish button would show a confirm modal,
  - Clicking the Finish button on the modal would close the modal and show the quiz result
    - If campers meet the passing grade
      - The quiz is disabled
      - The Finish button would change to "Submit and go"
    - If campers don't meet the passing grade
      - The quiz is disabled
      - The Finish button would be disabled

Ref https://github.com/freeCodeCamp/fCC10/issues/2#issuecomment-2391462203.

<!-- Feel free to add any additional description of changes below this line -->

huyenltnguyen committed:
feat(client): handle quiz finish and exit

huyenltnguyen committed:
chore: add e2e tests

moT01 commented on client/src/templates/Challenges/quiz/show.tsx:
```diff
@@ -139,17 +161,16 @@ const ShowQuiz = ({
-      openCompletionModal();
-      setIsPassed(true);
-    },
+    passingGrade: 80,
```
The quiz challenges all say "Answer all of the questions below correctly to pass the quiz." So those would need to be updated if we go this route. We haven't settled on anything. Let me get back to you on this.

Pretty minor, but the variable name could be changed to `passingPercent` - seems slightly more clear to me, but not necessary to change.

moT01 reviewed (changes requested):
This is looking good @huyenltnguyen 🎉 I mentioned below that we should set the passing percent to 85%. Other than that, I would be fine merging this. A couple other things I noticed...

This isn't using the space from the component library - so it might conflict with that other PR.

I don't get any feedback when I pass the quiz. Like a celebratory modal or anything.

No need to fix either of those here.

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into feat/quiz-modals

huyenltnguyen committed:
chore: migrate Spacer

huyenltnguyen committed:
fix: change passingGrade to 85

huyenltnguyen committed:
fix: add back the completion modal

moT01 reviewed (commented):
Yea, it's a little weird with the modal, but I think it's better - lets them know that they passed.

The spacer is still used in the two modal components.

huyenltnguyen committed:
ugh, spacer

ojeytonwilliams committed:
refactor: use useCallback to prevent effect firing

It's a bit more verbose, but means we can let the linter tell us which
deps we need. We don't have to think about it.

ojeytonwilliams commented:
Hey, @huyenltnguyen, I'll finish my review soon, but to explain: passing the deps to `usePageLeave` seemed like it would be a maintenance headache, since it required us to figure out exactly what `usePageLeave`'s effect really depended on. Wrapping the functions in `useCallback` means that the linter can find any missing deps and we don't have to.

ojeytonwilliams commented on client/src/templates/Challenges/quiz/exit-quiz-modal.tsx:
```diff
@@ -0,0 +1,66 @@
+  onExit: () => void;
+}
+
+const mapStateToProps = createSelector(
+  isExitQuizModalOpenSelector,
+  (isExitQuizModalOpen: boolean) => ({
+    isExitQuizModalOpen
+  })
+);
```
```suggestion
const mapStateToProps = (state: unknown) => ({
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
  isExitQuizModalOpen: isExitQuizModalOpenSelector(state)
});
```
We do this all over the place, but `createSelector` creates a memoized selector. If we're not computing anything, we don't need to memoize, it just creates an extra layer of abstraction.

ojeytonwilliams commented on client/src/templates/Challenges/quiz/exit-quiz-modal.tsx:
```diff
@@ -0,0 +1,66 @@
+  })
+);
+
+const mapDispatchToProps = (dispatch: Dispatch) =>
+  bindActionCreators(
+    {
+      closeExitQuizModal: () => closeModal('exitQuiz')
+    },
+    dispatch
+  );
```
```suggestion
const mapDispatchToProps = {
  closeExitQuizModal: () => closeModal('exitQuiz')
};
```
Same thing, but [less cruft](https://react-redux.js.org/using-react-redux/connect-mapdispatch#defining-mapdispatchtoprops-as-an-object).

ojeytonwilliams commented on client/src/templates/Challenges/quiz/exit-quiz-modal.tsx:
```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import { bindActionCreators, Dispatch } from 'redux';
+import { createSelector } from 'reselect';
```
```suggestion
```

ojeytonwilliams commented on client/src/templates/Challenges/quiz/finish-quiz-modal.tsx:
```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import { bindActionCreators, Dispatch } from 'redux';
+import { createSelector } from 'reselect';
```
```suggestion
```

ojeytonwilliams commented on client/src/templates/Challenges/quiz/finish-quiz-modal.tsx:
```diff
@@ -0,0 +1,67 @@
+  onFinish: () => void;
+}
+
+const mapStateToProps = createSelector(
+  isFinishQuizModalOpenSelector,
+  (isFinishQuizModalOpen: boolean) => ({
+    isFinishQuizModalOpen
+  })
+);
```
```suggestion
const mapStateToProps = (state: unknown) => ({
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
  isExitQuizModalOpen: isExitQuizModalOpenSelector(state)
});
```

ojeytonwilliams commented on client/src/templates/Challenges/quiz/finish-quiz-modal.tsx:
```diff
@@ -0,0 +1,67 @@
+  })
+);
+
+const mapDispatchToProps = (dispatch: Dispatch) =>
+  bindActionCreators(
+    {
+      closeFinishQuizModal: () => closeModal('finishQuiz')
+    },
+    dispatch
+  );
```
```suggestion
const mapDispatchToProps = {
  closeExitQuizModal: () => closeModal('exitQuiz')
};
```

ojeytonwilliams commented on e2e/quiz-challenge.spec.ts:
```diff
@@ -0,0 +1,139 @@
+import { test, expect } from '@playwright/test';
+
+test.describe('Quiz challenge', () => {
+  test.skip(
+    () => process.env.SHOW_UPCOMING_CHANGES !== 'true',
+    'The FED superblock is not available if SHOW_UPCOMING_CHANGES is false'
+  );
```
Just an aside, but this is effectively "skip(() => true)", since we never test upcoming changes in e2e tests.

At some point soonish it would be good if we could run tests against both, but I'm not sure how best to achieve that. We want to avoid just running all the tests again, but with `SHOW_UPCOMING_CHANGES=true`, since it just takes too long.

ojeytonwilliams replied:
I see what you're saying. We do need a way to test this stuff, otherwise it'll just be a source of false confidence. However, precisely how we should do that isn't obvious to me. Perhaps we need a separate set of tests that run against the upcoming build. It would probably mean wasting a bit more time in CI building both versions of the client, but once the new api is in we can get rid of a few workflows. We should be able to handle the extra build time.

ojeytonwilliams commented on e2e/quiz-challenge.spec.ts:
```diff
@@ -0,0 +1,139 @@
+  test('should show a confirm exit modal when user clicks on the exit button', async ({
+    page
+  }) => {
+    test.setTimeout(30000);
```
Why is this so slow?

huyenltnguyen committed:
Apply suggestions from code review

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into feat/quiz-modals

huyenltnguyen committed:
fix: finish modal props

huyenltnguyen committed:
fix: remove timeout extension and skip flaky test on chrome and safari

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into feat/quiz-modals

huyenltnguyen committed:
fix: update challenge URL

ojeytonwilliams committed:
refactor: avoid unnecessary state

unanswered was being used as a proxy for "should we show the unanswered
questions or not?", so I changed it to be more direct.

ojeytonwilliams reviewed (approved):
Hey @huyenltnguyen I pushed a small refactor, I hope that's okay.

Just one question: is it placeholder behaviour that the quiz locks after failure? If this is not intended to be the final UX, it's all good.

ojeytonwilliams commented:
Perfect, thanks for the clarification.
